File: ruby/medium/binary_tree_zigzag_level_order_traversal.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[][]}
def zigzag_level_order(root)
    queue = Array.new
    res, temp = Array.new, Array.new
    queue.push(root) if root
    current_nodes, num_nodes_in_level, level = 0, queue.length, 0
    while !queue.empty?
        current = queue.shift
        if((level % 2) == 0)
            temp.push(current.val)
        else
            temp.unshift(current.val)
        end
        current_nodes += 1
        queue.push(current.left) if current.left
        queue.push(current.right) if current.right
        if current_nodes == num_nodes_in_level
            res.push(temp)
            temp = Array.new
            num_nodes_in_level = queue.length
            current_nodes = 0
            level += 1
        end
    end
    return res
end

File: javascript/medium/binaryTreeZigzagLevelOrderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function(root) {
    const res = new Array(), queue = new Array();
    if(root === null) {
        return res;
    }
    queue.push(root);
    let numNodesInLevel = queue.length, currentCount = 0, level = 0, levelNodes = new Array();
    while(queue.length > 0) {
        let current = queue.shift();
        if((level % 2) === 0) {
            levelNodes.push(current.val);    
        } else {
            levelNodes.unshift(current.val);
        }
        currentCount++;
        if(current.left !== null) { queue.push(current.left); }
        if(current.right !== null) { queue.push(current.right); }
        
        if(currentCount === numNodesInLevel) {
            res.push(levelNodes);
            levelNodes = new Array();
            numNodesInLevel = queue.length;
            currentCount = 0;
            level++;
        }
    }
    
    return res;
};
